Add small dev grant blog post (#371)
sunpy/sunpy.org: posts/2023/2023-05-10-timeseries_sdg.md

```{post} May 10 2023
:author: David Stansby

```

# The future of time series data in `sunpy`

In late 2022 I got a [small development grant](https://numfocus.org/programs/small-development-grants) from [NumFocus](https://numfocus.org/) to scope the future of time series data in `sunpy`.

> The successful application [can be read on the sunpy wiki](https://github.com/sunpy/sunpy/wiki/2022-Timeseries-Small-Development-Grant). The application contains context that I won't repeat here.

Reviewer: ```suggestion
The successful application [can be read on the sunpy wiki](https://github.com/sunpy/sunpy/wiki/2022-Timeseries-Small-Development-Grant).
The application contains context that I won't repeat here.
```

This blog post is the key outcome of this grant, with a record of what I did, the recommendations I made, and any decisions we came to as a community.

## User requirements
The first stage of my work investigated what the user requirements are for a sunpy data container for time series data. As part of this I used my own experience and the following community engagement:
- Discussion at one of the weekly sunpy community meetings in December 2022
- [Discussion on the Python in Heliophysics mailing list](https://groups.google.com/g/pyhc-list/c/ujnZEZcsI_k/m/gNfYrIJdBgAJ)
- [Discussion on the SunPy forum](https://community.openastronomy.org/t/choosing-a-future-data-container-for-timeseries-data/556?u=dstansby)

> From these discsusion came the following list of requirements:

Reviewer: ```suggestion
From these discussion came the following list of requirements:
```

| Store multi-dimensional data | Although time is a common index to timeseries data, it isn't always the only one. As an example, velocity distribution functions measured in the solar wind are 4D datasets, with data as a function of time and three dimensions in velocity space. |
| Handle time scales with leapseconds | Some timescales can contain timestamps that occur within a leapsecond.  |
| Store and use physical units with the data and any non-time indices | |
| Store data in a format that can be used with scientific Python libraries | |
| Support for storing out-of memory datasets | |
| Store metadata alongside actual data | |
| Have a way to store an observer coordinate alongside the time index | |
| Have an easy way to do common data manipulation tasks | e.g. interpolating, resampling, rebinning |
| Have a way to combine multiple timeseries objects, and keep track of metadata | |
| Ability to convert to other common time series objects (e.g. `pandas.DataFrame`) | |
| Functionality for loading and saving out to common file formats | |

## Existing options for a data container

> The next step was to identify a set of possible data containers that could be used to store time- series data in sunpy. The identified options were:

Reviewer: ```suggestion
The next step was to identify a set of possible data containers that could be used to store time- series data in sunpy.
The identified options were:
```

| Package | Container |
| ------- | --------- |
| [sunpy](https://github.com/sunpy/sunpy/) | Custom `TimeSeries` object, backed by `pandas.DataFrame` |
| [HAPI Client](https://github.com/hapi-server/client-python) | `numpy.ndarray` |
| [pySPEDAS](https://github.com/spedas/pyspedas) | Not clear if users can access the data itself |
| [spacepy](https://github.com/spacepy/spacepy) | Unclear if there is any specific timeseries container object |
| [aidapy](https://gitlab.com/aidaspace/aidapy) | `xarray.DataArray`|
| [cdflib](https://github.com/MAVENSDC/cdflib) | `numpy.ndarray` |
| [NDCube](https://github.com/sunpy/ndcube) | `NDCube` |
| [pytplot](https://github.com/MAVENSDC/pytplot) | `xarray.DataArray` |
| [solo-epd-loader](https://github.com/jgieseler/solo-epd-loader) | `pandas.DataFrame` |
| [speasy](https://github.com/SciQLop/speasy) | Custom `DataContainer` object, backed by `numpy.ndarray` |

There is no common container used, with only `astropy.TimeSeries` not represented out of the possible options above.

### What datasets does sunpy currently support?

> sunpy currently has built in support for reading CDF files that conform to the [Space Physics Guidelines for CDF](https://spdf.gsfc.nasa.gov/sp_use_of_cdf.html), as long as the dataset is one- or two- dimensional. Alongside this several custom data readers have been written to support different data sources:

Reviewer: ```suggestion
sunpy currently has built in support for reading CDF files that conform to the [Space Physics Guidelines for CDF](https://spdf.gsfc.nasa.gov/sp_use_of_cdf.html), as long as the dataset is one- or two- dimensional.
Alongside this several custom data readers have been written to support different data sources:
```

### `astropy.timeseries.TimeSeries`
| | | |
| -- | -- | -- |
| Time-like index data | 🟩 | |
| Different time scales | 🟩 | |
| Multi-dimensional data | 🛑 | |
| Physical units | 🟩 | |
| Interop with scientific Python | 🟩 | |
| Out of memory | 🟠 | Apparently there is some support, but this is undocumented. |
| Metadata | 🟩  | Can store on the `.meta` attribute  |
| Observer coordinates | 🟩 |  |
| Easy data manipulation | 🟠 | |
| I/O | 🟩 | I/O is done via `astropy.table.Table` |

### `NDCube`
| | | |
| -- | -- | -- |
| Time-like index data | 🟩 | |
| Different time scales | 🟩 | |
| Multi-dimensional data | 🟩 | |
| Physical units | 🟩 |  |
| Interop with scientific Python | 🟩 | |
| Out of memory | 🟠 | Seems to be supported in theory, but little docs  |
| Metadata | 🟩  | Can store arbitrary FITS metadata |
| Observer coordinates | 🛑 | No support for extra coordinates |
| Easy data manipulation | 🛑 | Very few manipulation methods impelmented |
| I/O | 🛑 | |

## Initial recommendations
- `numpy.ndarray` doesn't implement several key features, and these are almost certainly out of scope for future `ndarray` development, so I suggest `ndarray` is discounted.

> - `xarray.DataArray` builds on top of `pandas.DataFrame` with additional features that would be useful to us, I suggest `pandas.DataFrame` is dicsounted.

Reviewer: ```suggestion
- `xarray.DataArray` builds on top of `pandas.DataFrame` with additional features that would be useful to us, I suggest `pandas.DataFrame` is discounted.
```

- `NDCube` is designed specifically to store data that is associated with a FITS world coordinate system (WCS). While some solar timeseries data is already in the FITS format, a large portion is in CDF format which is tabular, which FITS is not primarily designed to represent. So I suggest `NDCube` is discounted.

At a SunPy community meeting there was a consensus agreement that going forward we should consider `astropy.TimeSeries` and `xarray.DataArray` as the two options to consider.

These two options have the following comparison:

| | `astropy.TimeSeries` | `xarray.DataArray`|
| -- | -- | -- |
| Time-like index data | 🟩 | 🟩 |
| Different time scales | 🟩 | 🛑 |
| Multi-dimensional data | 🛑 | 🟩 |
| Physical units | 🟩 | 🛑 |
| Interop with scientific Python | 🟩 | 🟩 |
| Out of memory | 🟠  | 🟩 |
| Metadata | 🟩  | 🟩  |

> | Observer coordinates | 🛑 | 🟠 |

Reviewer 2: This dosen't match up with the section on astropy above where observer coordinates (correctly?) has a green mark.

| Easy data manipulation | 🟠 | 🟩 |
| I/O | 🟩 | 🟩 |

> My initial recommendation would be to adopt `xarray.DataArray`, as there are more green items compared to `astropy.TimeSeries`. I also think the two red items have the possibility of being solved with `DataArray`:

Reviewer: ```suggestion
My initial recommendation would be to adopt `xarray.DataArray`, as there are more green items compared to `astropy.TimeSeries`.
I also think the two red items have the possibility of being solved with `DataArray`:
```

> My initial recommendation would be to adopt `xarray.DataArray`, as there are more green items compared to `astropy.TimeSeries`. I also think the two red items have the possibility of being solved with `DataArray`:

Reviewer 2: I also think that this is missing any discussion on `SkyCoord` etc compatibility with xarray (which personally is my biggest worry).

Author: 👍 fair - I've added a not that I didn't have time to investigate this.

> My initial recommendation would be to adopt `xarray.DataArray`, as there are more green items compared to `astropy.TimeSeries`. I also think the two red items have the possibility of being solved with `DataArray`:

Author: Fixed

> - It *should* (I haven't confirmed this) be possible to convert times in different time scales (including ones with leapseconds) to a single timescale that doesn't have leapseconds, and store this in an `xarray.DataArray`.

Reviewer: ```suggestion
- It *should* (I haven't confirmed this) be possible to convert times in different time scales (including ones with leap seconds) to a single timescale that doesn't have leap seconds, and store this in an `xarray.DataArray`.
```

- Although there is not native support for units in `DataArray` currently, there is interest and ongoing development to support them.

Finally, `xarray` has a much bigger development community than `astropy.TimeSeries`, so implementing bug fixes and new features would probably be much easier with `xarray`.

## Putting `astropy` objects in `xarray` structures

For the final part of the small development grant, I investigated the changes needed to put `astropy` objects in `xarray` structures.

> As a model for doing this, it is currently possible to store unitful data created with `pint` in `xarray` structures. Support for doing this has two components:

Reviewer: ```suggestion
As a model for doing this, it is currently possible to store unitful data created with `pint` in `xarray` structures.
Support for doing this has two components:
```

- `xarray` [natively supports storing duck arrays](https://docs.xarray.dev/en/stable/internals/duck-arrays-integration.html)

> - `xarray-pint` provides a set of accessors that can be used to serialise and de-serialise unitful data so that it can be saved to a file and loaded again. It does this by converting the unit data into metadata, with strings representing units.

Reviewer: ```suggestion
- `xarray-pint` provides a set of accessors that can be used to serialise and deserialise unitful data so that it can be saved to a file and loaded again.
  It does this by converting the unit data into metadata, with strings representing units.
```

> It is not currently possible to store `astropy.Quantity` objects in `xarray` structures, as they inherit directly from `ndarray`, and get co-erced from `Quantity` to `ndarray` durying the `xarray` structure initialisation. I think fixing this is (at least initially) a one line change, changing [this line](https://github.com/pydata/xarray/blob/51554f2638bc9e4a527492136fe6f54584ffa75d/xarray/core/variable.py#L288) from

Reviewer: ```suggestion
It is not currently possible to store `astropy.Quantity` objects in `xarray` structures, as they inherit directly from `ndarray`, and get coerced from `Quantity` to `ndarray` during the `xarray` structure initialisation.
I think fixing this is (at least initially) a one line change, changing [this line](https://github.com/pydata/xarray/blob/51554f2638bc9e4a527492136fe6f54584ffa75d/xarray/core/variable.py#L288) from
```

```python
data = np.asarray(data)
```
to
```python
if not isinstance(data, np.array):

>     data = np.asarray(data)

Reviewer 2: Can't you go one easier here and use `np.asanyarray` which allows all subclasses through?

Author: 🤷

```

> Before moving forward with this it needs to be possible to run the full unit tests in xarray with `astropy.Quantity`. I started work on this in these two PRs:

Reviewer: ```suggestion
Before moving forward with this it needs to be possible to run the full unit tests in xarray with `astropy.Quantity`.
I started work on this in these two PRs:
```
